### PR TITLE
Add placeholder prop to supporting form inputs (#446)

### PR DESCRIPTION
**Pull request by justin-m-morgan**

### CODE REVIEW

File: lib/surface/components/form/email_input.ex [hunk 1]
```diff
@@ -21,8 +21,11 @@ defmodule Surface.Components.Form.EmailInput do
   import Surface.Components.Utils, only: [events_to_opts: 1]
   import Surface.Components.Form.Utils
 
+  @doc "Placeholder text."
```
[the remark this answers is not in the document]
Comment (justin-m-morgan) [reply lib/surface/components/form/email_input.ex]: Thank you for the suggestion.  I made the above recommended changes to all the affected files.

File: lib/surface/components/form/telephone_input.ex [hunk 2]
```diff
@@ -24,8 +24,11 @@ defmodule Surface.Components.Form.TelephoneInput do
   @doc "A regular expression to validate the entered value"
   prop pattern, :string
 
+  @doc "Placeholder text."
+  prop placeholder, :string
+
   def render(assigns) do
-    helper_opts = props_to_opts(assigns)
+    helper_opts = props_to_opts(assigns, [:placeholder])
```
[the remark this answers is not in the document]
Comment (miguel-s) [reply lib/surface/components/form/telephone_input.ex]: Yep, you got it right 👍

**Comment by miguel-s** [comment]
@justin-m-morgan Thanks for the PR ❤️ 

>  if we are going to that direction, I believe we should go further by implementing all the available attributes depending on the input: https://www.w3schools.com/tags/tag_input.asp WDYT?

@Malian IMO that would add a lot of nice DX compared to only having the `opts` prop. I remember that @msaraiva had some reservations about this approach; iirc his concern was around having bloated components, but I can't find our conversation. @msaraiva do I remeber correctly?

**Comment by justin-m-morgan** [comment]
> if we are going to that direction, I believe we should go further by implementing all the available attributes depending on the input: https://www.w3schools.com/tags/tag_input.asp WDYT?

I'm available to do this work if it's decided that is the direction you want to you.

**Comment by msaraiva** [comment]
> @Malian IMO that would add a lot of nice DX compared to only having the `opts` prop. I remember that @msaraiva had some reservations about this approach; iirc his concern was around having bloated components, but I can't find our conversation. @msaraiva do I remember correctly?

The main issue is about user expectations. If we start defining all available attributes as props to some components, we'll have to do it all of them, otherwise the users will always be in doubt if an attribute is defined as a prop or not, making the general usage of components less intuitive. The way we try to overcome that is by defining a simple easy-to-remember rule. Currently, the rule is that we only define a prop for an existing HTML attribute if:

1. The attribute is specific to the related HTML element. For instance, the `<RangeInput>` defines `min`, `max` and `step` as those are not the generic global attributes.
2. We need to add some extra behaviour for that attribute, e.g. add validation (type check or if it's required), define a default value, etc.

Those a the most basic rules we could come up with and, although it's not perfect, it's still much more intuitive them randomly adding props to some components. What we could do, however, as middle ground solution for form controls, is to redefine the first statement as something like: 

1. The attribute is specific to the related HTML element or a **category of elements (e.g. inputs)**. For instance, the `<RangeInput>` defines `min`, `max` and `step` (as those are not the generic global attributes) and the `value` attribute, which is defined for every input.

In this case, I'm not sure which category the `placeholder` attribute belongs to. Is it available for all inputs? If not, can we identify all components that have that attribute as a homogeneous group (category)? Like text-based inputs or whatever.

If so, I think we can certainly move on in that direction and add not only `placeholder` but all attributes belonging to that same identifiable group, otherwise, we should probably wait a bit until we introduce the concept of "interfaces", which will make it easier to define/share groups of props declarations.

**Comment by justin-m-morgan** [comment]
@miguel-s  @msaraiva @Malian Did we decide on this issue or can we merge the current changes?

**Comment by Malian** [comment]
@justin-m-morgan We haven't decided what we should do.

@msaraiva `placeholder` is an attribute that can be used with input and with textarea, but not all kind of inputs... The simplest counter example is the `radio` input that does not accept a placeholder.

My point of view is that, I expect to have all the available attributes for any kind of Form/HTML element. I find it strange to be able to use some attributes but not the others.

That being said, I share Marlus point of view about "all or nothing":

> If we start defining all available attributes as props to some components, we'll have to do it all of them, otherwise the users will always be in doubt if an attribute is defined as a prop or not, making the general usage of components less intuitive.

The main disadvantage of this solution is that we should be up-to-date with the HTML elements. Who will check that aspect?

All the allowed attributes are available here: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input

In conclusion, I suggest to not merge this PR. Here are two possibles next step:
1) Close the issue, and stay with the solution we have
2) Add all the available attributes to `Form` components.

@miguel-s @msaraiva @justin-m-morgan WDYT?
